File: apollo-android-support/src/main/java/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheFactory.java
package com.apollographql.apollo.cache.normalized.sql;

import com.apollographql.apollo.cache.normalized.NormalizedCacheFactory;
import com.apollographql.apollo.cache.normalized.RecordFieldJsonAdapter;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

public final class SqlNormalizedCacheFactory extends NormalizedCacheFactory<SqlNormalizedCache> {
  private final ApolloSqlHelper helper;

  public SqlNormalizedCacheFactory(ApolloSqlHelper helper) {
    this.helper = checkNotNull(helper, "helper == null");
  }

  @Override public SqlNormalizedCache create(RecordFieldJsonAdapter recordFieldAdapter) {
    return new SqlNormalizedCache(recordFieldAdapter, helper);
  }
}

File: apollo-normalized-cache-sqlite/src/main/java/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheFactory.kt
package com.apollographql.apollo.cache.normalized.sql

import android.content.Context
import com.apollographql.apollo.cache.normalized.NormalizedCacheFactory
import com.apollographql.apollo.cache.normalized.RecordFieldJsonAdapter
import com.squareup.sqldelight.android.AndroidSqliteDriver
import com.squareup.sqldelight.db.SqlDriver

class SqlNormalizedCacheFactory internal constructor(
    driver: SqlDriver
) : NormalizedCacheFactory<SqlNormalizedCache>() {

  @JvmOverloads
  constructor(context: Context, name: String? = "apollo.db")
      : this(AndroidSqliteDriver(ApolloDatabase.Schema, context.applicationContext, name))

  constructor(apolloSqlHelper: ApolloSqlHelper) : this(apolloSqlHelper.sqlDriver)

  private val apolloDatabase = ApolloDatabase(driver)

  override fun create(recordFieldAdapter: RecordFieldJsonAdapter) =
      SqlNormalizedCache(recordFieldAdapter, apolloDatabase.cacheQueries)

}
